src/trio/_repl.py

```diff
@@ -0,0 +1,114 @@
+from __future__ import annotations
+
+import ast
+import contextlib
+import inspect
+import sys
+import types
+import warnings
+from code import InteractiveConsole
+from typing import Generator
+
+import trio
+import trio.lowlevel
+
+if sys.version_info < (3, 11):
+    from exceptiongroup import BaseExceptionGroup
+
+
+def _flatten_exception_group(
+    excgroup: BaseExceptionGroup[BaseException],
+) -> Generator[BaseException, None, None]:
+    for exc in excgroup.exceptions:
+        if isinstance(exc, BaseExceptionGroup):
+            yield from _flatten_exception_group(exc)
+        else:
+            yield exc
+
+
+class TrioInteractiveConsole(InteractiveConsole):
+    # code.InteractiveInterpreter defines locals as Mapping[str, Any]
+    # but when we pass this to FunctionType it expects a dict. So
+    # we make the type more specific on our subclass
+    locals: dict[str, object]
+
+    def __init__(self, repl_locals: dict[str, object] | None = None):
+        super().__init__(locals=repl_locals)
+        self.compile.compiler.flags |= ast.PyCF_ALLOW_TOP_LEVEL_AWAIT
+
+    def runcode(self, code: types.CodeType) -> None:
+        async def _runcode_in_trio() -> BaseException | None:
+            func = types.FunctionType(code, self.locals)
+            try:
+                coro = func()
+            except BaseException as e:
+                return e
+
+            if inspect.iscoroutine(coro):
+                try:
+                    await coro
+                except BaseException as e:
+                    return e
```

Comment:
Can some of this exception stuff be replaced with stuff from `outcome`? I don't see it providing much technical benefit but it would decrease some code duplication.

Reply:
I might not be understanding how outcomes works, but I don't see how it would simplify this. Wouldn't it just end up  calling capture method and then immediately unwrapping?
```python
try:
  coro = outcome.capture(func)
  coro.unwrap()
except BaseException as e:
  return e
```

The traceback printing is consistent with the standard python repl and the `python -m asyncio` so, I'm leaning towards keeping it as is.

Reply:
Well specifically I was thinking directly returning the Value/Error (and an `if` on whether to use `capture` or `acapture`? Which would disallow leaving out the `await` when calling an async function but that's already trio style. I'll make this a separate comment when I get back to looking at this). If the trace backs are the same then the only benefit would be a few less lines, which probably isn't worth the work.

Reply:
I came up with this. It's better in some ways, but at the same time bit strange.
```python
    def runcode(self, code: types.CodeType) -> None:
        async def _runcode_in_trio():
            func = types.FunctionType(code, self.locals)
            result = outcome.capture(func)
            
            if isinstance(result, outcome.Error):
                return result
            
            coro = result.unwrap()
            if inspect.iscoroutine(coro):
                return await outcome.acapture(lambda: coro)
        
            return outcome.Value(coro)

        try:
            value = trio.from_thread.run(_runcode_in_trio).unwrap()
        except SystemExit:
            # If it is SystemExit quit the repl. Otherwise, print the
            # traceback.
            # There could be a SystemExit inside a BaseExceptionGroup. If
            # that happens, it probably isn't the user trying to quit the
            # repl, but an error in the code. So we print the exception
            # and stay in the repl.
            raise
        except BaseException:
            self.showtraceback()
```
I think it is awkward because there is a single step for most expressions, but any repl input that is an `await` needs the second step of evaluation.

Reply:
I finally tried poking at this a bit, ultimately what I was thinking about is:

```py
    def runcode(self, code: types.CodeType) -> None:
        async def _runcode_in_trio():
            func = types.FunctionType(code, self.locals)
            if inspect.iscoroutinefunction(func):
              return outcome.acapture(func)
            else:
              return outcome.capture(func)

        try:
            value = trio.from_thread.run(_runcode_in_trio).unwrap()
        except SystemExit:
            # If it is SystemExit quit the repl. Otherwise, print the
            # traceback.
            # There could be a SystemExit inside a BaseExceptionGroup. If
            # that happens, it probably isn't the user trying to quit the
            # repl, but an error in the code. So we print the exception
            # and stay in the repl.
            raise
        except BaseException:
            self.showtraceback()
```

probably doesn't work cause I haven't tested exactly that, but `inspect.iscoroutinefunction` seems to do the right thing (doesn't trigger for `trio.sleep(1)` or `async def x(): ...`, but triggers for `await trio.sleep(1)`)

Reply:
Ah, `iscoroutinefunction` is the bit I was missing. That is much better. Thanks for the nudge in that direction. I've got a few bits to double check but it looks like that works right.

Reply:
Normally it might not be the best idea (since it can’t detect code that just returns a coroutine object directly). But in this case we know `return` isn’t valid.